### PR TITLE
Implement Blender-like pan mode

### LINKED ISSUE #70: Support Panning of the Editor Camera
# Summary
Enable panning of the camera in the editor viewport. `Shift + Right Mouse` seems like a reasonable command.

## Notes

- Change will need to be made in [EditorCameraController.java](https://github.com/Interrupt/delverengine/blob/33a71d0ebe66fb31fa94784a22885674be44bb5c/DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java#L61)
- Vertical mouse movement should move camera along the `camera.up` vector.
- Horizontal mouse movement should move along the camera's right vector. Which should be `camera.direction` cross product `camera.up`

### PR DESCRIPTION
**Pull request by evrimoztamur**
Enables pan mode in the editor by Shift + Right click dragging.

Resolves #70, refer to lines 180-197 and 246. Auto-formatting led to severe spacing issues due to mix.

### CODE REVIEW

File: DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java [hunk 1]
```diff
@@ -12,30 +12,33 @@
 import com.interrupt.helpers.AnimationHelper;
 import com.interrupt.helpers.InterpolationHelper;
 
-/** Subsystem for controlling and positioning the editor camera. */
+/**
+ * Subsystem for controlling and positioning the editor camera.
```
Comment (joshuaskelly) [inline DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java:16]:
```java
/** Single line JavaDocs please. :) */
```

File: DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java [hunk 2]
```diff
@@ -63,124 +66,132 @@ public void tick() {
...
+            Vector3 pivotPosition = new Vector3(camera.direction).scl(orbitDistance).add(camera.position);
+
+            // Calculate new camera position.
+            cameraNewDirection.scl(-orbitDistance);
+            cameraNewDirection.add(pivotPosition);
+
+            position.set(cameraNewDirection.x, cameraNewDirection.z, cameraNewDirection.y);
+        }
+
+        if (Gdx.input.isKeyPressed(Input.Keys.SHIFT_LEFT) && Gdx.input.isButtonPressed(Input.Buttons.RIGHT)) { // Pan mode enabled (SHIFT_LEFT held)
```
Comment (joshuaskelly) [inline DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java:180]: Whenever I see boolean logic with a little note next to it, my instinct is to wrap that up in a method. Maybe something like:
```java
if (isPanning()) { ... }
```

File: DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java [hunk 3]
```diff
@@ -203,41 +214,20 @@ public void setRotation(float x, float y) {
...
-                if (!Gdx.input.isCursorCatched()) {
-                    Gdx.input.setCursorCatched(true);
-                }
+            if (!Gdx.input.isCursorCatched()) {
+                Gdx.input.setCursorCatched(true);
             }
 
-            rotation.x += moveX * 0.005f;
-            rotation.y -= moveY * 0.005f;
+            if (!Gdx.input.isKeyPressed(Input.Keys.SHIFT_LEFT)) { // Normal rotation, pan mode disabled (SHIFT_LEFT released)
```
Comment (joshuaskelly) [inline DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java:225]: Same here:
```java
if (isRotating()) { ... }
```

File: DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java [hunk 4]
```diff
@@ -248,48 +238,52 @@ public boolean scrolled(int amount) {
         return false;
     }
 
-    /** Position the camera to view the current selection. */
+    /**
+     * Position the camera to view the current selection.
```
Comment (joshuaskelly) [inline DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java:242]: +1 JavaDoc

File: DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java [hunk 5]
```diff
@@ -248,48 +238,52 @@ public boolean scrolled(int amount) {
...
+            bounds.getCenter(selectedPosition);
+        }
+
+        Vector3 cameraOffset = new Vector3(camera.direction.x, camera.direction.z, camera.direction.y).scl(orbitDistance);
+        Vector3 finalPosition = new Vector3(selectedPosition).sub(cameraOffset);
+        moveTo(finalPosition);
+    }
+
+    /**
+     * Smoothly move to given destination.
```
Comment (joshuaskelly) [inline DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java:277]: +1 JavaDoc

**Review by joshuaskelly — APPROVED** [review]
Some little nitpicks, but it's working well for me!

File: DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java (hunk 2 above)
Comment (evrimoztamur) [reply DelvEdit/src/com/interrupt/dungeoneer/editor/EditorCameraController.java:180]: I intended to unwrap these while refactoring the controller code and making these checks/states more clear, actually. I can amend this with my clean-up if that's better?

**Comment by evrimoztamur** [comment]
Refactored a large portion of the movement logic, adjusting behaviour where necessary. Let me know if there are questions on the specific changes.

**Comment by evrimoztamur** [comment]
Issue was caused by the adjustment of `orbitDistance` after the `zMovement` was corrected in the movement code. Led to the value being shrunk by `cos(rotation.y)` at each move and deviating further from the actual distance to the pivot.

**Review by joshuaskelly — APPROVED** [review]
Looks good to me! 👍

**Pull request merged**